File: app/src/main/java/com/canchapp_kotlin/ui/auth/AuthViewModel.kt
package com.canchapp_kotlin.ui.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.canchapp_kotlin.data.network.LoginResponse
import com.canchapp_kotlin.data.network.RegisterResponse
import com.canchapp_kotlin.data.network.UserDto
import com.canchapp_kotlin.data.repository.AuthRepository
import com.canchapp_kotlin.util.Resource
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class AuthViewModel : ViewModel() {

    private val repository = AuthRepository()

    private val _loginState = MutableStateFlow<Resource<LoginResponse>?>(null)
    val loginState = _loginState.asStateFlow()

    private val _registerState = MutableStateFlow<Resource<RegisterResponse>?>(null)
    val registerState = _registerState.asStateFlow()

    private val _loggedUser = MutableStateFlow<UserDto?>(null)
    val loggedUser = _loggedUser.asStateFlow()

    fun login(identifier: String, password: String) {
        viewModelScope.launch {
            _loginState.value = Resource.Loading
            val result = repository.login(identifier, password)
            _loginState.value = result
            if (result is Resource.Success) {
                _loggedUser.value = result.data.data?.user
            }
        }
    }

    fun register(
        email: String,
        firstName: String,
        lastName: String,
        password: String,
        roleName: String,
        username: String
    ) {
        viewModelScope.launch {
            _registerState.value = Resource.Loading
            _registerState.value = repository.register(email, firstName, lastName, password, roleName, username)
        }
    }

    fun logout() {
        _loggedUser.value = null
        _loginState.value = null
    }

    fun resetLoginState() { _loginState.value = null }
    fun resetRegisterState() { _registerState.value = null }
}

File: ios/CanchApp/ViewModels/AuthViewModel.swift
import Foundation

/// Equivalent of Kotlin's AuthViewModel.
/// Uses @MainActor so @Published updates always happen on the main thread.
@MainActor
final class AuthViewModel: ObservableObject {

    @Published var loginState:    Resource<LoginResponse>?    = nil
    @Published var registerState: Resource<RegisterResponse>? = nil
    @Published var loggedUser:    UserDto?                    = nil

    private let repository = AuthRepository()

    // MARK: - Login

    func login(identifier: String, password: String) {
        loginState = .loading
        Task {
            let result = await repository.login(identifier: identifier, password: password)
            loginState = result
            if case .success(let response) = result {
                loggedUser = response.data?.user
            }
        }
    }

    // MARK: - Register

    func register(
        email: String,
        firstName: String,
        lastName: String,
        password: String,
        roleName: String,
        username: String
    ) {
        registerState = .loading
        Task {
            registerState = await repository.register(
                email: email, firstName: firstName, lastName: lastName,
                password: password, roleName: roleName, username: username
            )
        }
    }

    // MARK: - Logout

    func logout() {
        loggedUser    = nil
        loginState    = nil
        registerState = nil
    }

    func resetLoginState()    { loginState    = nil }
    func resetRegisterState() { registerState = nil }
}
